src/gcc_fix/missing_gcc_functions.c

```diff
@@ -35,177 +37,212 @@ void* memset(void* str, s32 c, size_t n) {
+    if ((s64)c < 0) {
+        // cvt.s.l assumes signed input, adjust output
+        v += 4294967296.0f; // 2^32
+    }
+    return v;
 }
 
-f64 __floatundidf(u32 c) {
-    return (f64)c;
+// u64 -> f64
+f64 __floatundidf(u64 c) {
+    register union {
+        f64 f;
+        u64 i;
+    } m;
+    register f64 v;
+
+    m.i = c;
+    __asm__ ("cvt.d.l %0, %1" : "=f"(v) : "f"(m.f));
+    if ((s64)c < 0) {
+        // cvt.d.l assumes signed input, adjust output
+        v += 18446744073709551616.0; // 2^64
+    }
+    return v;
 }
 
-f32 __powisf2(f32 a, s32 b) {
-    const s32 recip = b < 0;
-    f32 r = 1;
+// Compute x^m by binary exponentiation
 
-    while (1) {
-        if (b & 1) {
-            r *= a;
-        }
+f32 __powisf2(f32 x, s32 m) {
+    u32 n = (m < 0) ? -m : m;
+    f32 y = (n % 2 != 0) ? x : 1.0f;
 
-        b /= 2;
+    while (n >>= 1) {
+        x = x * x;
 
-        if (b == 0) {
-            break;
+        if (n % 2 != 0) {
+            y = y * x;
         }
-
-        a *= a;
     }
-    return recip ? 1 / r : r;
+    return (m < 0) ? (1.0f / y) : y;
 }
 
+// Compute division and modulo of 64-bit signed and unsigned integers
+
 __asm__("                                   \n\
     .set push                               \n\
-    .set noat                               \n\
     .set noreorder                          \n\
     .set gp=64                              \n\
                                             \n\
 .global __umoddi3                           \n\
 __umoddi3:                                  \n\
-    .type __umoddi3, @function              \n\
-    .ent __umoddi3                          \n\
-    sw    $a0, ($sp)                        \n\
-    sw    $a1, 4($sp)                       \n\
-    sw    $a2, 8($sp)                       \n\
-    sw    $a3, 0xc($sp)                     \n\
-    ld    $t7, 8($sp)                       \n\
-    ld    $t6, ($sp)                        \n\
-    ddivu $zero, $t6, $t7                   \n\
-    bnez  $t7, 1f                           \n\
-     nop                                    \n\
-    break 7                                 \n\
-1:                                          \n\
-    mfhi  $v0                               \n\
-    dsll32 $v1, $v0, 0                      \n\
-    dsra32 $v1, $v1, 0                      \n\
-    jr    $ra                               \n\
+  .type __umoddi3, @function                \n\
+  .ent __umoddi3                            \n\
+    sw      $a0, 0x0($sp)                   \n\
+    sw      $a1, 0x4($sp)                   \n\
+    sw      $a2, 0x8($sp)                   \n\
+    sw      $a3, 0xC($sp)                   \n\
+    ld      $t6, 0($sp)                     \n\
+    ld      $t7, 8($sp)                     \n\
+    dremu   $v0, $t6, $t7                   \n\
+    dsll32  $v1, $v0, 0                     \n\
+    dsra32  $v1, $v1, 0                     \n\
+    jr      $ra                             \n\
```

Comment:
Since you are using inline asm for the other functions, wouldn't it be better to rewrite this functions as C functions with inline asm too? or it isn't worth it?

Reply:
Explained in discord but for visibility: these functions aren't worth rewriting in inline asm due to the o32 abi passing values in 2 registers. The inputs/outputs for inline asm aren't really possible to use in this case and it becomes about as much manual work as these pure-asm functions.